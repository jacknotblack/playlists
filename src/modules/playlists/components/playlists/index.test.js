import React from "react";
import { shallow } from "enzyme";
import PlayLists from "./index";
import {
  playListswithSongDetailFilled as playLists,
  songs
} from "../../testData";

const setup = props => {
  const enzymeWrapper = shallow(<PlayLists {...props} />);
  return {
    props,
    enzymeWrapper
  };
};

describe("<PlayLists />", () => {
  const initProps = {
    playLists,
    songs,
    getInitPlayLists: jest.fn(),
    addList: jest.fn()
  };
  describe("render", () => {
    const { enzymeWrapper, props } = setup(initProps);

    it("renders without crashing", () => {
      expect(enzymeWrapper.length).toBe(1);
    });
    describe("componentDidMount", () => {
      // const { enzymeWrapper, props } = setup(initProps);
      it('should trigger "getInitPlayLists"', () => {
        enzymeWrapper.instance().componentDidMount();
        expect(props.getInitPlayLists).toHaveBeenCalled();
      });
    });
    describe("render playList", () => {
      describe("playList length = 0", () => {
        it("should match snapshot", () => {
          const { enzymeWrapper } = setup({
            ...initProps,
            playLists: []
          });
          expect(enzymeWrapper).toMatchSnapshot();
        });
      });

      describe("playList is undefined", () => {
        const { enzymeWrapper } = setup({ ...initProps, playLists: undefined });
        it("should match snapshot", () => {
          expect(enzymeWrapper).toMatchSnapshot();
        });
      });
    });
  });

  describe("function props", () => {
    describe("handle input", () => {
      it("should change input value by chaging state", () => {
        const { enzymeWrapper, props } = setup(initProps);
        enzymeWrapper.setState({
          listNameInput: "My uber list"
        });
        expect(enzymeWrapper).toMatchSnapshot();
      });
      it("should trigger setState when input value changed", () => {
        const { enzymeWrapper, props } = setup(initProps);
        expect(enzymeWrapper.state().listNameInput).toBe("");
        const event = {
          target: {
            value: "I got a new name!"
          }
        };
        enzymeWrapper.find(".add-list > input").simulate("change", event);
        expect(enzymeWrapper.state().listNameInput).toBe("I got a new name!");
      });
    });
    it("should alert and do nothing else when add button is clicked and input value is empty", () => {
      const { enzymeWrapper, props } = setup(initProps);
      enzymeWrapper
        .find(".add-list > button")
        .get(0)
        .props.onClick();
      expect(window.alert).toHaveBeenCalledWith(
        "you can't add a playlist without a name"
      );
      expect(props.addList).not.toHaveBeenCalled();
    });

    it("should alert and do nothing else when add button is clicked and input value matches exist playlist name", () => {
      const { enzymeWrapper, props } = setup(initProps);
      enzymeWrapper.setState({ listNameInput: "abczzz" });
      enzymeWrapper
        .find(".add-list > button")
        .get(0)
        .props.onClick();
      expect(window.alert).toHaveBeenCalledWith("abczzz already exists");
      expect(props.addList).not.toHaveBeenCalled();
    });

    it('should trigger "addList" and setState when add button is clicked', () => {
      const { enzymeWrapper, props } = setup(initProps);
      enzymeWrapper.setState({ listNameInput: "super unique name" });
      enzymeWrapper
        .find(".add-list > button")
        .get(0)
        .props.onClick();
      expect(window.alert).not.toHaveBeenCalled();
      expect(props.addList).toHaveBeenCalledWith("super unique name");
      expect(enzymeWrapper.state().listNameInput).toBe("");
    });
  });
});
