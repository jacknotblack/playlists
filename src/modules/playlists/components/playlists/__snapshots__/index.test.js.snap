// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayLists /> function props handle input should change input value by chaging state 1`] = `
<div
  className="playlists"
>
  <div
    className="add-list"
  >
    <input
      onChange={[Function]}
      type="text"
      value="My uber list"
    />
    <button
      onClick={[Function]}
    >
      Add List
    </button>
  </div>
  <div
    className="lists"
  >
    <PlayList
      key="0"
      name="aaa"
      songs={
        Array [
          Object {
            "artist": "Jack",
            "duration": "1:28",
            "id": 0,
            "title": "abc1",
          },
          Object {
            "artist": "Randy",
            "duration": "5:41",
            "id": 3,
            "title": "abc4",
          },
        ]
      }
    />
    <PlayList
      key="1"
      name="bbb"
      songs={Array []}
    />
    <PlayList
      key="2"
      name="ccc"
      songs={
        Array [
          Object {
            "artist": "Tyler",
            "duration": "2:33",
            "id": 2,
            "title": "abc3",
          },
        ]
      }
    />
    <PlayList
      key="3"
      name="ddd"
      songs={
        Array [
          Object {
            "artist": "Randy",
            "duration": "5:41",
            "id": 3,
            "title": "abc4",
          },
        ]
      }
    />
    <PlayList
      key="4"
      name="zzz"
      songs={Array []}
    />
    <PlayList
      key="5"
      name="zzzz"
      songs={Array []}
    />
    <PlayList
      key="6"
      name="abc"
      songs={Array []}
    />
    <PlayList
      key="7"
      name="abczzz"
      songs={Array []}
    />
    <PlayList
      key="8"
      name="123"
      songs={Array []}
    />
  </div>
</div>
`;

exports[`<PlayLists /> render render playList playList is undefined should match snapshot 1`] = `
<div
  className="playlists"
>
  <div
    className="add-list"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Add List
    </button>
  </div>
  <div
    className="lists"
  >
    <span>
      No playlist yet
    </span>
  </div>
</div>
`;

exports[`<PlayLists /> render render playList playList length = 0 should match snapshot 1`] = `
<div
  className="playlists"
>
  <div
    className="add-list"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
    >
      Add List
    </button>
  </div>
  <div
    className="lists"
  >
    <span>
      No playlist yet
    </span>
  </div>
</div>
`;
